pkg/kube/teams_test.go

```diff
@@ -0,0 +1,108 @@
+package kube
+
+import (
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"testing"
+
+	kube_mocks "k8s.io/client-go/kubernetes/fake"
+)
+
+func TestGetAdminNamespace(t *testing.T) {
+	t.Parallel()
+	namespace := `apiVersion: v1
+kind: Namespace
+metadata:
+  annotations:
+    jenkins-x.io/created-by: Jenkins X
+    jenkins-x.io/admin-namespace: admin-namespace
+  labels:
+    env: dev
+    team: myteam
+  name: myteam
+status:
+  phase: Active`
+
+	client, err := createMockClient(namespace)
+	require.NoError(t, err, "failed to create mock client")
+	adminNamespace, err := GetAdminNamespace(client, "myteam")
+	assert.NoError(t, err, "GetAdminNamespace should not error")
+	assert.Equal(t, "admin-namespace", adminNamespace)
+}
+
+func TestGetAdminNamespaceDefaulted(t *testing.T) {
+	t.Parallel()
+	namespace := `apiVersion: v1
+kind: Namespace
+metadata:
+  annotations:
+    jenkins-x.io/created-by: Jenkins X
+  labels:
+    env: dev
+    team: myteam
+  name: myteam
+status:
+  phase: Active`
+
+	client, err := createMockClient(namespace)
+	require.NoError(t, err, "failed to create mock client")
+	adminNamespace, err := GetAdminNamespace(client, "myteam")
+	assert.NoError(t, err, "error occurred calling GetAdminNamespace")
+	assert.Equal(t, "myteam", adminNamespace)
+}
+
+// TestGetAdminNamespaceOnNonTeam tests that when there are no annotations the code does not blow up.
+// there should always be one annotation "jenkins-x.io/created-by: Jenkins X" for a team but better to not blow up
+func TestGetAdminNamespaceOnNonTeam(t *testing.T) {
+
+	t.Parallel()
+	namespace := `apiVersion: v1
+kind: Namespace
+metadata:
+  name: myteam
+status:
+  phase: Active`
+
+	client, err := createMockClient(namespace)
+	require.NoError(t, err, "failed to create mock client")
+	adminNamespace, err := GetAdminNamespace(client, "myteam")
+	assert.NoError(t, err, "error occurred calling GetAdminNamespace")
+	assert.Equal(t, "myteam", adminNamespace)
+}
+
+func TestSetAdminNamespace(t *testing.T) {
+	t.Parallel()
+	namespace := `apiVersion: v1
+kind: Namespace
+metadata:
+  annotations:
+    jenkins-x.io/created-by: Jenkins X
+  labels:
+    env: dev
+    team: myteam
+  name: myteam
+status:
+  phase: Active`
+	client, err := createMockClient(namespace)
+	require.NoError(t, err, "failed to create mock client")
+	err = SetAdminNamespace(client, "myteam", "my-admin-namespace")
+	assert.NoError(t, err, "error occurred calling SetAdminNamespace")
+	adminNs, err := GetAdminNamespace(client, "myteam")
+	assert.NoError(t, err, "error occurred calling GetAdminNamespace")
+	assert.Equal(t, "my-admin-namespace" ,adminNs)
+}
+
+func createMockClient(objDef string) (kubernetes.Interface, error) {
+	decode := scheme.Codecs.UniversalDeserializer().Decode
+	obj, _, err := decode([]byte(objDef), nil, nil)
```

Comment:
Why not to construct directly the namespace object instead of mashaling YAML? If someone, modifies the YAML content by mistake the tests will fail.

Reply:
in constructing the object I can make the same mistake twice (as I am querying the object later), so doing it this way means I can catch those mistakes.
As for if someone modified the yaml and the tests fail then that is no different to them changing an object and causing the test to fail - the yaml is part of the contract of the test so I do not understand that comment.

Reply:
What about upgrading the kuberntes client library, which might introduce an incompatibility in the schema which is less likely but yeah. IMHO is a bit fragile.

Reply:
k8s can't just change the schema it will break all the things,  everytones deployment.yaml etc etc the world will fall down (and if they do change it it would be a new version) otherwise it will break and everything.
the internal object representaion is more likely to change than the interchange format  - so I think to say this is more fragile is actually incorrect - it is less fragile.